[WIP]Short term solution for capital income MTR

Amy-Xu opened the pull request:
A new function for capital income MTR is added in this PR due to the following two reasons.

- The one-cent margin needs to be weighted split among all capital income sources before calculation. This step is only needed for these capital income sources but not any other type of income.

- Not much validation needed at this point since we only calculate mtr based on a certain set of capital income sources.

This is still just a short-term solution, as it's easy to implement and relatively efficient computation wise, without messing up the original mtr function which is currently used by other team members for testing purposes. In long run, we need to combine the two mtr functions. 

@MattHJensen @martinholmer Could you take a look at this function? I haven't added any test yet, but will do once you're happy with the function.

MattHJensen commented on taxcalc/calculate.py:
```diff
@@ -275,6 +275,48 @@ def mtr(self, income_type_str='e00200p',
         return (mtr_fica, mtr_iit, mtr_combined)
 
+    def capital_mtr(self):
+        capital_income_sources = ['e00300', 'e00400', 'e00600', 'e00650',
```
It might be easier to test for consistency against the other mtr function (until we merge the two) if this list were a default argument. Then you could have a couple of kinds of tests. 

1. pass `['e00300']` to this function and make sure you get the same as when you use the preexisting `mtr()`. 

2. pass `['e00300', 'e2350']` to this function and make sure that you get the same thing as if you calculate the mtrs separately for those variables via the preexisting `mtr()` and then weight the mtrs after the fact.

Amy-Xu replied:
Got you. Will do.

Amy-Xu commented:
@jdebacker In this PR you can see the new function I added for capital income sources. It includes 10 types of income you mentioned and split the one-cent margin according to each income type in proportion to the total capital income of that taxpayer. For taxpayers without any capital income, they get zero margin added to their capital income variables and therefore end up with zero MTR. Is that what you expect? Could you take a look at this function as well? Please let me know if anything doesn't make sense.

feenberg commented:
On Wed, 2 Dec 2015, Amy Xu wrote:

> 
> @jdebacker In this PR you can see the new function I added for capital income sources.
> It includes 10 types of income you mentioned and split the one-cent margin according
> to each income type in proportion to the total capital income of that taxpayer. For
> taxpayers without any capital income, they get zero margin added to their capital
> income variables and therefore end up with zero MTR. Is that what you expect? Could
> you take a look at this function as well? Please let me know if anything doesn't make
> sense.

I don't think this is going to be right, but I am not sure of the details.

If you split the penny in proportion to capital income shares, there is a 
problem if there in no capital income - the formula would have you 
dividing by zero, since the share of each form of capital income is that 
amount of income divided by the total capital income. It is not right to 
substitute zero for the undefined result. Consider the effect, which is to 
mix a bunch of zeroes in for taxpayers who would in fact pay tax on 
capital income.

If what we want is the average marginal rate in the income clasess, the 
the right thing to do is to calculate the tax, then multiply every capital 
income item by 1.01 and recalculate the tax. Then divide the aggregate 
change in tax by the aggregate change in income. This gives the average 
marginal rate of tax on capital income (which is not the same as the 
average of the taxpayer marginal rates).

If what we want is a rate to use in a behavorial equation, then we need a 
rate for taxpayers with no capital income and we should spread the penny 
accross the capital income items in some plausible but arbitrary manner. 
For example, in proportion to the average across taxpayers with such 
income.

If this isn't clear, we should have a phone call.

dan
617-863-0343

> 
> —
> Reply to this email directly or view it on
> GitHub.[AHvQVW3URZVuC1ZxN97SD2zKAdSXr5svks5pL0uogaJpZM4GsY9S.gif]
> 
> 
>

Amy-Xu commented:
Just spoke with Dan, here's a quick update. Among all the issues Dan pointed out, only one thing left is the mtr function for taxpayers without any capital income. It's incorrect to simply give them zero marginal tax rates, which is what I'm doing right now. We need to find a way to split the one-cent margin properly. This would make a pretty huge difference. @MattHJensen @jdebacker FYI

jdebacker commented:
Excellent. Dan's suggestions are great.

Re the zero capital income filers - I check the PUF and filers with capital income in [0, $100) have an average of 95% of their capital income from interest income.  

Shall we just round and assume that additional capital income for those with none comes entirely from interest income?  This looks to be a pretty good approximation.

@Amy-Xu

Amy-Xu commented:
Ahhhhh that's exactly what Dan suggested. Forgot to add that in my note. Will do asap.

Sent from my iPhone

On Dec 2, 2015, at 5:18 PM, Jason DeBacker <notifications@github.com<mailto:notifications@github.com>> wrote:


Excellent. Dan's suggestions are great.

Re the zero capital income filers - I check the PUF and filers with capital income in [0, $100) have an average of 95% of their capital income from interest income.

Shall we just round and assume that additional capital income for those with none comes entirely from interest income? This looks to be a pretty good approximation.

@Amy-Xu<https://github.com/Amy-Xu>

—
Reply to this email directly or view it on GitHub<https://github.com/open-source-economics/Tax-Calculator/pull/492#issuecomment-161451817>.

feenberg commented:
On Wed, 2 Dec 2015, Jason DeBacker wrote:

> 
> Excellent. Dan's suggestions are great.
> 
> Re the zero capital income filers - I check the PUF and filers with capital income in
> [0, $100) have an average of 95% of their capital income from interest income.
> 
> Shall we just round and assume that additional capital income for those with none
> comes entirely from interest income? This looks to be a pretty good approximation.

I agree.

dan

> 
> @Amy-Xu
> 
> —
> Reply to this email directly or view it on
> GitHub.[AHvQVRzDK74DV2UjKMNC3N4-zKX45KWUks5pL2W2gaJpZM4GsY9S.gif]
> 
> 
>

Amy-Xu commented:
@jdebacker @feenberg How should we split the margin if the taxpayer's different income sources cancelled out in the sum? I'm testing the scenario where includes two capital income, e00300 and e23250, in the MTR function. One taxpayer happens to have 173 for e00300 and -173 for e23250. The sum of the two is zero and I added an epsilon to the sum, which makes the split super bizarre, but I'm not sure how to solve it either.

Amy-Xu commented:
An update on the difference between two methods, one splits the margin and calculates mtr and another calculates mtr for each item first and weighted sum in the end. Out of all five records showing difference, four of them are due to the fact that they fall on the kinks of capital tax rate function. Specifically, after stacking their capital gains and dividends on their ordinary income, you can see that the total (c24530) sits at the second to the lowest tax bracket, which separates zero and 15% percent rate for capital gains. The last record, even thought I'm not 100% sure at this point, I guess the diff is from the issue I post above. Will update again on this.

jdebacker commented:
I think in general you want to think about how to compute the MTR across
all components if any of the components in negative.  E.g. if the component
is negative do we want to multiply it by 1.01?  That doesn't seem quite
right.  It seems like you'd want to multiply by -1.01 in that case.  But I
haven't thought about this too much and am happy to defer to @feenberg on
this.



On Thu, Dec 3, 2015 at 10:08 AM, Amy Xu <notifications@github.com> wrote:

> An update on the difference between two methods, one splits the margin and
> calculates mtr and another calculates mtr for each item first and weighted
> sum in the end. Out of all five records showing difference, four of them
> are due to the fact that they fall on the kinks of capital tax rate
> function. Specifically, after stacking their capital gains and dividends on
> their ordinary income, you can see that the total (c24530) sits at the
> second to the lowest tax bracket, which separates zero and 15% percent rate
> for capital gains. The last record, even thought I'm not 100% sure at this
> point, I guess the diff is from the issue I post above. Will update again
> on this.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/open-source-economics/Tax-Calculator/pull/492#issuecomment-161692277>
> .
>

Amy-Xu commented:
@jdebacker yea good call! Definitely haven't thought about all negative values yet. But let me explain this first, I also explained to Dan yesterday but didn't summarize it here. Right now we are not adding the margin by multiplying income by 1.01, which would matter if we average mtrs across all taxpayers. Instead, we simply added one cent to their income. This should be equivalent to the multiply method in our case.

feenberg commented:
Unless the kink produces a wild rate, no worries.

dan

On Thu, 3 Dec 2015, Amy Xu wrote:

> 
> An update on the difference between two methods, one splits the margin and calculates mtr and another
> calculates mtr for each item first and weighted sum in the end. Out of all five records showing difference,
> four of them are due to the fact that they fall on the kinks of capital tax rate function. Specifically,
> after stacking their capital gains and dividends on their ordinary income, you can see that the total
> (c24530) sits at the second to the lowest tax bracket, which separates zero and 15% percent rate for
> capital gains. The last record, even thought I'm not 100% sure at this point, I guess the diff is from the
> issue I post above. Will update again on this.
> 
> —
> Reply to this email directly or view it on
> GitHub.[AHvQVdTL5bzddohuGlDms4JJpIlr_THWks5pMGB0gaJpZM4GsY9S.gif]
> 
> 
>

Amy-Xu commented:
@feenberg The kinks indeed produces wild rates for some records, like 10e+22 level. I don't think that's normal for MTRs.

Another important issue worth your attention is how we should deal with negative capital income in general. The case I ran into was two kinds of capital income cancelled each other, which makes it difficult to split the margin before MTR calculation. How do you think we should deal with that?

feenberg commented:
On Thu, 3 Dec 2015, Amy Xu wrote:

> 
> @feenberg The kinks indeed produces wild rates for some records, like 10e+22 level. I
> don't think that's normal for MTRs.

You get that number from adding only a penny? That would be a tax 
difference of 10e20 dollars, which is pretty big. Can you find where in 
the code the notch occurs? Maybe the taxpayer has a million dollars of one 
type of income and only a dollar of another. Then the finite difference 
for the smaller amount is only a millionth of a cent and a small notch 
could give a large MTR. Often a notch indicates a programming error. If 
you look at the intermediate results, it may be clear where something big 
and wrong occurs. Legislative notches are typically only a few dollars.

> 
> Another important issue worth your attention is how we should deal with negative
> capital income in general. The case I ran into was two kinds of capital income
> cancelled each other, which makes it difficult to split the margin before MTR
> calculation. How do you think we should deal with that?

The macro guys should say what they need, but I wouldn't calculate the 
rates separately for each income type if it causes trouble as mentioned 
above. For a taxpayer with losses, adding a penny to income does produce 
the correct measure of the wedge in capital returns for that taxpayer.

dan

> 
> —
> Reply to this email directly or view it on
> GitHub.[AHvQVUlSFeJLeJK7lVIHj-5bDbwG1O_1ks5pMJUTgaJpZM4GsY9S.gif]
> 
> 
>

jdebacker commented:
@Amy-Xu Ok - then let's just do what you are doing for negative income (adding a positive amount) and use the implied MTRs even changes cancel out - this shouldn't result in cancelations, right?  If so, they seem legit and we are ok with MTR = 0 in those cases.

Re the large MTRs.  Until you can determine the cause and whether it's due to what Dan suggests, we can still proceed with what you have - we "clean" the data anyway - dropping really large ATR/MTRs from the data before we fit our functions to the data.

Amy-Xu commented:
Just figured out where the problem is for the extremely large MTRs. They overlap with those records that have two income variables canceled out. The reason why they end up with unreasonably large value is that I added an epsilon (10e-20) to the sum, which's supposed to be zero for those records. And then used the sum to calculate ratios of each income source in the sum for margin split. So the ratios ended up extremely large. But I figure this problem probably only exist when we use two variables as input in the testing scenarios. I'll see how likely this is going to happen when we use all ten types of capital income.

feenberg commented:
On Thu, 3 Dec 2015, Jason DeBacker wrote:

> 
> @Amy-Xu Ok - then let's just do what you are doing for negative income (adding a
> positive amount) and use the implied MTRs even changes cancel out - this shouldn't
> result in cancelations, right? If so, they seem legit and we are ok with MTR = 0 in
> those cases.
> 
> Re the large MTRs. Until you can determine the cause and whether it's due to what Dan
> suggests, we can still proceed with what you have - we "clean" the data anyway -
> dropping really large ATR/MTRs from the data before we fit our functions to the data.

plus or minus 100% would be a reasonable MTR filter. Labor income is often 
-.45 MTR (because of the EITC), and there are possibly other interactions 
that might make for even more extreme rates, so don't exclude negative 
rates.

dan

> 
> —
> Reply to this email directly or view it on
> GitHub.[AHvQVT_tnJ95nXamjYCGfus57PG6uDzrks5pMLbWgaJpZM4GsY9S.gif]
> 
> 
>

jdebacker commented:
Thanks, Dan!  Will do.

On Thu, Dec 3, 2015 at 4:48 PM, feenberg <notifications@github.com> wrote:

>
>
> On Thu, 3 Dec 2015, Jason DeBacker wrote:
>
> >
> > @Amy-Xu Ok - then let's just do what you are doing for negative income
> (adding a
> > positive amount) and use the implied MTRs even changes cancel out - this
> shouldn't
> > result in cancelations, right? If so, they seem legit and we are ok with
> MTR = 0 in
> > those cases.
> >
> > Re the large MTRs. Until you can determine the cause and whether it's
> due to what Dan
> > suggests, we can still proceed with what you have - we "clean" the data
> anyway -
> > dropping really large ATR/MTRs from the data before we fit our functions
> to the data.
>
> plus or minus 100% would be a reasonable MTR filter. Labor income is often
> -.45 MTR (because of the EITC), and there are possibly other interactions
> that might make for even more extreme rates, so don't exclude negative
> rates.
>
> dan
>
> >
> > —
> > Reply to this email directly or view it on
> > GitHub.[AHvQVT_tnJ95nXamjYCGfus57PG6uDzrks5pMLbWgaJpZM4GsY9S.gif]
>
> >
> >
> >
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/open-source-economics/Tax-Calculator/pull/492#issuecomment-161813390>
> .
>

Amy-Xu commented:
Closing, replace by #497
